pkg/sources/forward/data_forward.go

```diff
@@ -270,50 +323,52 @@ func (df *DataForward) forwardAChunk(ctx context.Context) {
-	// a read message to the corresponding write message
-	transformerResults := make([]readWriteMessagePair, len(readMessages))
-	// applyTransformer, if there is an Internal error, it is a blocking call and
-	// will return only if shutdown has been initiated.
-
-	// create a pool of Transformer Processors
-	var wg sync.WaitGroup
-	for i := 0; i < df.opts.transformerConcurrency; i++ {
-		wg.Add(1)
-		go func() {
-			defer wg.Done()
-			df.concurrentApplyTransformer(ctx, transformerCh)
-		}()
-	}
-	concurrentTransformerProcessingStart := time.Now()
+	readWriteMessagePairs := make([]readWriteMessagePair, len(readMessages))
+
+	// If a user defined transformer exists, apply it
+	if df.transformer != nil {
+		// user-defined transformer concurrent processing request channel
+		transformerCh := make(chan *readWriteMessagePair)
+
+		// create a pool of Transformer Processors
+		var wg sync.WaitGroup
+		for i := 0; i < df.opts.transformerConcurrency; i++ {
+			wg.Add(1)
+			go func() {
+				defer wg.Done()
+				df.concurrentApplyTransformer(ctx, transformerCh)
+			}()
+		}
+		concurrentTransformerProcessingStart := time.Now()
+		processReadMessages(true, readMessages, processorWM, readWriteMessagePairs, df, transformerCh)
 
-	// send to transformer only the data messages
-	for idx, m := range readMessages {
-		// emit message size metric
-		metrics.ReadBytesCount.With(map[string]string{metrics.LabelVertex: df.vertexName, metrics.LabelPipeline: df.pipelineName, metrics.LabelVertexType: string(dfv1.VertexTypeSource), metrics.LabelVertexReplicaIndex: strconv.Itoa(int(df.vertexReplica)), metrics.LabelPartitionName: df.reader.GetName()}).Add(float64(len(m.Payload)))
-		// assign watermark to the message
-		m.Watermark = time.Time(processorWM)
-		// send transformer processing work to the channel
-		transformerResults[idx].readMessage = m
-		transformerCh <- &transformerResults[idx]
+		// let the go routines know that there is no more work
+		close(transformerCh)
+		// wait till the processing is done. this will not be an infinite wait because the transformer processing will exit if
+		// context.Done() is closed.
+		wg.Wait()
+		df.opts.logger.Debugw("concurrent applyTransformer completed",
+			zap.Int("concurrency", df.opts.transformerConcurrency),
+			zap.Duration("took", time.Since(concurrentTransformerProcessingStart)),
+		)
+
+		metrics.SourceTransformerConcurrentProcessingTime.With(map[string]string{
+			metrics.LabelVertex:             df.vertexName,
+			metrics.LabelPipeline:           df.pipelineName,
+			metrics.LabelVertexReplicaIndex: strconv.Itoa(int(df.vertexReplica)),
+			metrics.LabelPartitionName:      df.reader.GetName(),
+		}).Observe(float64(time.Since(concurrentTransformerProcessingStart).Microseconds()))
+		// transformer processing is done.
+	} else {
+		processReadMessages(false, readMessages, processorWM, readWriteMessagePairs, df, nil)
```

Comment:
This function call has `hasUDTransformer` as `false` and `tranformerCh` as `nil`. I sense an opportunity for improvement. Can we include line L328->L364 in the processReadMessages method such that we can call `processReadMessages(df.transformer != nil, readMessages, processorWM, readWriteMessagePairs, df)` and inside of `processReadMessages`, conditionally create a transformerCh?

Two benefits: 1. more readable code as developers might not care about the transformer when looking through the code. 2. de-coupled input parameters of `processReadMessages`.

Reply:
I refactored a bit and if we create the channel inside the helper function we would essentially just move the entire logic into the function, so i decided to just keep it in the forwardAChunk method and make a helper for emitting the metrics.